File: web/src/ga/AdaptiveGA.ts
// 增强自适应遗传算法
import { GeneticAlgorithm } from './GeneticAlgorithm';
import { Maze } from '../maze/Maze';
import { Individual } from './Individual';
import type { GAConfig } from '../types';

export class AdaptiveGeneticAlgorithm extends GeneticAlgorithm {
  private initialMutationRate: number;
  private initialCrossoverRate: number;
  private initialPopulationSize: number;
  private diversityHistory: number[] = [];
  private stagnationCounter: number = 0;
  private lastBestFitness: number = 0;

  constructor(maze: Maze, config: GAConfig) {
    super(maze, config);
    this.initialMutationRate = config.mutationRate;
    this.initialCrossoverRate = config.crossoverRate;
    this.initialPopulationSize = config.populationSize;
  }

  evolve() {
    // 自适应调整参数
    if (this.generation > 0) {
      this.adaptParameters();
    }
    return super.evolve();
  }

  private adaptParameters(): void {
    const diversity = this.calculateDiversity();
    this.diversityHistory.push(diversity);

    const progress = this.generation / this.config.maxGenerations;
    const currentBest = this.bestIndividual?.fitness || 0;

    // 检测停滞
    if (Math.abs(currentBest - this.lastBestFitness) < 0.1) {
      this.stagnationCounter++;
    } else {
      this.stagnationCounter = 0;
    }
    this.lastBestFitness = currentBest;

    if (this.config.adaptiveMode === 'diversity') {
      this.adaptByDiversity(diversity);
    } else if (this.config.adaptiveMode === 'hybrid') {
      this.adaptHybrid(diversity, progress);
    }

    // 自适应种群大小
    this.adaptPopulationSize(progress);
  }

  /**
   * 基于多样性的自适应策略
   */
  private adaptByDiversity(diversity: number): void {
    if (this.diversityHistory.length > 10) {
      const avgDiversity = this.diversityHistory.slice(-10).reduce((a, b) => a + b) / 10;
      
      // 多样性过低：增加变异率，降低交叉率
      if (diversity < avgDiversity * 0.5) {
        this.config.mutationRate = Math.min(0.4, this.config.mutationRate * 1.2);
        this.config.crossoverRate = Math.max(0.5, this.config.crossoverRate * 0.95);
      } 
      // 多样性过高：降低变异率，增加交叉率
      else if (diversity > avgDiversity * 1.5) {
        this.config.mutationRate = Math.max(0.01, this.config.mutationRate * 0.9);
        this.config.crossoverRate = Math.min(0.95, this.config.crossoverRate * 1.05);
      }
    }
  }

  /**
   * 混合自适应策略：结合进化进度、多样性、停滞检测
   */
  private adaptHybrid(diversity: number, progress: number): void {
    // 1. 基于进化进度的基础调整
    const baseMutation = this.initialMutationRate * (1 - progress * 0.6);
    const baseCrossover = this.initialCrossoverRate * (1 + progress * 0.15);

    // 2. 根据停滞情况调整
    if (this.stagnationCounter > 20) {
      // 长期停滞：大幅提高变异率
      this.config.mutationRate = Math.min(0.5, baseMutation * 2.0);
      this.config.crossoverRate = Math.max(0.5, baseCrossover * 0.8);
      
      // 注入随机个体
      if (this.stagnationCounter > 30 && this.stagnationCounter % 10 === 0) {
        this.injectRandomIndividuals(0.1); // 注入10%随机个体
      }
    } else if (this.stagnationCounter > 10) {
      // 轻微停滞：适度调整
      this.config.mutationRate = Math.min(0.3, baseMutation * 1.3);
    } else {
      // 正常进化
      this.config.mutationRate = baseMutation;
      this.config.crossoverRate = baseCrossover;
    }

    // 3. 根据多样性微调
    if (diversity < 10 && this.fitnessHistory.length > 10) {
      const recentImprovement = this.fitnessHistory[this.fitnessHistory.length - 1].best - 
                               this.fitnessHistory[this.fitnessHistory.length - 10].best;
      if (recentImprovement < 5) {
        this.config.mutationRate = Math.min(0.5, this.config.mutationRate * 1.2);
      }
    }

    // 确保参数在合理范围内
    this.config.mutationRate = Math.max(0.01, Math.min(0.5, this.config.mutationRate));
    this.config.crossoverRate = Math.max(0.5, Math.min(0.95, this.config.crossoverRate));
  }

  /**
   * 自适应种群大小
   */
  private adaptPopulationSize(progress: number): void {
    // 后期收敛时缩减种群以节省计算
    if (progress > 0.7 && this.bestIndividual?.reachedEnd) {
      const targetSize = Math.max(
        Math.floor(this.initialPopulationSize * 0.6),
        this.config.elitismCount * 3
      );
      
      if (this.population.length > targetSize) {
        // 保留精英和部分优秀个体
        const sorted = [...this.population].sort((a, b) => b.fitness - a.fitness);
        this.population = sorted.slice(0, targetSize);
        this.config.populationSize = targetSize;
      }
    }
  }

  /**
   * 注入随机个体以增加多样性
   */
  private injectRandomIndividuals(ratio: number): void {
    const injectCount = Math.floor(this.population.length * ratio);
    const sorted = [...this.population].sort((a, b) => a.fitness - b.fitness);
    
    for (let i = 0; i < injectCount; i++) {
      sorted[i] = new Individual(this.maze, [], this.config.maxSteps);
    }
    
    this.population = sorted;
    this.stagnationCounter = 0; // 重置停滞计数
  }

  /**
   * 自适应选择压力
   */
  protected selection(): Individual {
    // 根据进化阶段调整锦标赛大小
    const progress = this.generation / this.config.maxGenerations;
    const tournamentSize = Math.floor(3 + progress * 4); // 3到7之间
    
    const tournament: Individual[] = [];
    
    for (let i = 0; i < tournamentSize; i++) {
      const idx = Math.floor(Math.random() * this.population.length);
      tournament.push(this.population[idx]);
    }
    
    return tournament.reduce((best, ind) => 
      ind.fitness > best.fitness ? ind : best
    );
  }
}

File: src/adaptive_ga.py
"""
自适应遗传算法模块 - 动态调整参数
"""
import numpy as np
from typing import List
from genetic_algorithm import GeneticAlgorithm, Individual
from maze import Maze


class AdaptiveGeneticAlgorithm(GeneticAlgorithm):
    """自适应遗传算法 - 根据进化状态动态调整参数"""
    
    def __init__(self, maze: Maze, 
                 population_size: int = 100,
                 max_generations: int = 500,
                 initial_mutation_rate: float = 0.3,
                 initial_crossover_rate: float = 0.7,
                 elitism_count: int = 2,
                 max_steps: int = 200,
                 adaptive_mode: str = 'diversity'):
        """
        初始化自适应遗传算法
        
        Args:
            adaptive_mode: 自适应模式
                - 'diversity': 基于种群多样性调整
                - 'fitness': 基于适应度改进调整
                - 'hybrid': 混合模式
        """
        super().__init__(maze, population_size, max_generations, 
                        initial_mutation_rate, initial_crossover_rate,
                        elitism_count, max_steps)
        
        self.initial_mutation_rate = initial_mutation_rate
        self.initial_crossover_rate = initial_crossover_rate
        self.adaptive_mode = adaptive_mode
        
        # 自适应参数历史
        self.mutation_history = []
        self.crossover_history = []
        self.diversity_history = []
        
    def calculate_diversity(self) -> float:
        """计算种群多样性（适应度标准差）"""
        if not self.population:
            return 0.0
        fitness_values = [ind.fitness for ind in self.population]
        return float(np.std(fitness_values))
    
    def calculate_improvement_rate(self) -> float:
        """计算适应度改进率"""
        if len(self.fitness_history) < 2:
            return 1.0
        
        recent = min(10, len(self.fitness_history))
        recent_best = [h['best_fitness'] for h in self.fitness_history[-recent:]]
        
        if len(recent_best) < 2:
            return 1.0
        
        improvement = recent_best[-1] - recent_best[0]
        return max(0.0, min(1.0, improvement / (recent_best[0] + 1)))
    
    def adapt_parameters(self):
        """根据种群状态自适应调整参数"""
        diversity = self.calculate_diversity()
        self.diversity_history.append(diversity)
        
        if self.adaptive_mode == 'diversity':
            # 基于多样性调整
            # 多样性低 → 增加变异率，促进探索
            # 多样性高 → 降低变异率，促进收敛
            if len(self.diversity_history) > 1:
                avg_diversity = np.mean(self.diversity_history[-10:])
                
                if diversity < avg_diversity * 0.5:  # 多样性过低
                    self.mutation_rate = min(0.4, self.mutation_rate * 1.2)
                    self.crossover_rate = max(0.5, self.crossover_rate * 0.95)
                elif diversity > avg_diversity * 1.5:  # 多样性过高
                    self.mutation_rate = max(0.05, self.mutation_rate * 0.9)
                    self.crossover_rate = min(0.9, self.crossover_rate * 1.05)
        
        elif self.adaptive_mode == 'fitness':
            # 基于适应度改进调整
            improvement_rate = self.calculate_improvement_rate()
            
            if improvement_rate < 0.01:  # 停滞不前
                self.mutation_rate = min(0.5, self.mutation_rate * 1.3)
            elif improvement_rate > 0.1:  # 快速改进
                self.mutation_rate = max(0.05, self.mutation_rate * 0.9)
        
        elif self.adaptive_mode == 'hybrid':
            # 混合模式：综合考虑
            diversity = self.calculate_diversity()
            improvement_rate = self.calculate_improvement_rate()
            progress = self.generation / self.max_generations
            
            # 初期：高变异率促进探索
            # 后期：低变异率促进收敛
            base_mutation = self.initial_mutation_rate * (1 - progress * 0.7)
            
            # 根据多样性和改进率微调
            if diversity < 10 and improvement_rate < 0.01:
                self.mutation_rate = min(0.5, base_mutation * 1.5)
            else:
                self.mutation_rate = max(0.05, base_mutation)
            
            # 交叉率随进度略微调整
            self.crossover_rate = self.initial_crossover_rate * (1 + progress * 0.2)
        
        # 记录参数变化
        self.mutation_history.append(self.mutation_rate)
        self.crossover_history.append(self.crossover_rate)
    
    def evolve(self) -> Individual:
        """执行一代进化（带自适应）"""
        # 先自适应调整参数
        if self.generation > 0:
            self.adapt_parameters()
        
        # 调用父类的evolve方法
        return super().evolve()
    
    def run(self, verbose: bool = True) -> Individual:
        """运行自适应遗传算法"""
        self.initialize_population()
        
        if verbose:
            print(f"自适应遗传算法 (模式: {self.adaptive_mode})")
            print(f"初始变异率: {self.initial_mutation_rate}")
            print(f"初始交叉率: {self.initial_crossover_rate}\n")
        
        for gen in range(self.max_generations):
            best = self.evolve()
            
            if verbose and (gen % 50 == 0 or best.reached_end):
                status = "已到达终点!" if best.reached_end else "未到达终点"
                print(f"代数 {gen}: 适应度={best.fitness:.2f}, "
                      f"路径长度={len(best.path)}, {status} | "
                      f"变异率={self.mutation_rate:.3f}, "
                      f"交叉率={self.crossover_rate:.3f}")
            
            if best.reached_end and gen > 50:
                if verbose:
                    print(f"\n在第 {gen} 代找到解决方案!")
                break
        
        if verbose:
            print(f"\n最终结果:")
            print(f"最佳适应度: {self.best_individual.fitness:.2f}")
            print(f"路径长度: {len(self.best_individual.path)}")
            print(f"最终变异率: {self.mutation_rate:.3f}")
            print(f"最终交叉率: {self.crossover_rate:.3f}")
        
        return self.best_individual

